spring-integration-event/src/main/java/org/springframework/integration/event/inbound/ApplicationEventListeningMessageProducer.java

```diff
@@ -116,20 +129,45 @@ public void onApplicationEvent(ApplicationEvent event) {
 		}
 	}
 
+	@Override
+	protected Object evaluatePayloadExpression(Object rootObject) {
+		if (!this.payloadExpressionSet && rootObject instanceof PayloadApplicationEvent) {
+			return ((PayloadApplicationEvent) rootObject).getPayload();
```

Comment:
You don't need to change anything for that, the event listener infrastructure is smart enough to extract what needs to be extracted. 

To be clear, if you emit `Message<PayloadApplicationEvent<Foo>>>` and that message is then pass to something that publish the event with the payload (that is `PayloadApplicationEvent<Foo>`), the receiver part can declare the method signature with any format. In particular this would work:

```
@EventListener
public void  handleFoo(Foo foo) { ....} 
```